### PR TITLE
Reaaaaaaaact: Add barebones new frontend in React

### PR DESCRIPTION
**Pull request by danielbeardsley**
See RFC at #199

This adds everything we need to get started recreating the frontend in
react / typescript / bootstrap / styled-components:

* New webpack config
* Easy offline dev with webpack-dev-server
* React-bootstrap
* React-typescript
* Component to connect to socket.io and get the pulls
* Available on the main app at /frontend
* TS Types for the Pull object and its components
* styled-components all ready and set up
* Keeps the old UI around and functional

### CODE REVIEW

**Comment by copperwall** [comment]
👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀

**Comment by kwiens** [comment]
What do you think about using [Preact](https://preactjs.com/) here to kick the tires with it? It should be a drop-in replacement, and save >100KB on load.

**Comment by ardelato** [comment]
Not really sure how to setup Pulldasher locally or how to run the `frontend:start` script on Cominor so at the moment any suggestions on how to properly setup the `/frontend` side of Pulldasher will be awesome. 

To note, it also seems that `process.env.BACKEND_URL` returns undefined in the file `frontend/src/page-context.ts`. Since in the console it logs a 404 error when fetching for the page context using the BACKEND_URL. 

CleanShot_Google Chrome app-Pulldasher

### Steps Taken:
1. Git cloned Pulldasher to my Cominor Home directory.
2. Ran `update-pulldasher`
3. Went to https://pulldasher.cominor.com/frontend/ 
4. Only saw the titles for all the Divs:

CleanShot_Google Chrome app-Pulldasher 2

**Comment by emmalopezcode** [comment]
CR 🎈

**Comment by addison-grant** [comment]
@danielbeardsley, should this be dev_blocked (more commits definintely incoming) or if it's ready for more QA/CR?

**Comment by addison-grant** [comment]
And any input on @ardelato's comment would be appreciated https://github.com/iFixit/pulldasher/pull/200#issuecomment-752793567

**Review by andyg0808 — APPROVED** [review]

**Comment by andyg0808** [comment]
Looks nice! 
CR :+1: through 733c455

File: frontend/src/pulldasher.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import { Pull } from './types';
+import Navbar from './navbar';
+import CRColumn from './cr-column';
+import Container from 'react-bootstrap/Container';
+import Row from 'react-bootstrap/Row';
+import Col from 'react-bootstrap/Col';
+
+export default function() {
+   const pulls: Pull[] = usePullsState();
+   return (<PullsContext.Provider value={{pulls:pulls}}>
```
[the remark this answers is not in the document]
Comment (danielbeardsley) [reply frontend/src/pulldasher.tsx:13]: I anticipate components all up and down the component stack will want the pulls collection and be able to access it without having to pass it down through multiple layers.

File: frontend/src/socket.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import getPageContext from './page-context';
+import * as io from 'socket.io-client';
+
+/**
+ * Giving socket.io some types
+ */
+export interface Socket {
+   on: (event: string, callback: (data: any) => void) => void;
+   emit: (event: string, data: any) => void;
+}
```
[the remark this answers is not in the document]
Comment (danielbeardsley) [reply frontend/src/socket.ts:10]: Sounds great! I doubt we're on the version that has those types and I didn't want to have this pull be *upgrade all the things* as well as introduce react.

**Comment by danielbeardsley** [comment]
Sorry @ardelato ! forgot I didn't reply yet here.

Pulldasher can now be run locally (well this UI anyway).

Clone the repo locally then:
1. `npm install`
2. `npm run frontend:start` (should open a browser with an in-progress version of pulldasher like below)

image

**Comment by andyg0808** [comment]
CR :+1: through b06c00b

**Comment by ardelato** [comment]
deploy_block❓ Confirmation on current working behavior.  Should pulls-whether fake or from GitHub-appear on the React Based Pulldasher?

@danielbeardsley I was able to get Pulldasher working locally, but it seems that the pulls from the `dummy-pulls.json` are not being populated on the frontend. 

In addition, there still seems to be a socket WebSocket issue on `Cominor`.

## Steps Taken:
### Local Pulldasher:
1. Ran `git clone git@github.com:iFixit/pulldasher.git` on my local computer
2. `cd` into the directory and ran `git checkout reaaaaaaaaaaact`
3. Ran `npm install` and it successfully completed the `postinstall` script:  "webpack && npm run frontend:build"
4. Ran `npm run frontend:start` and was able to load Pulldasher on localhost
5. Did not see any `pulls` from the `dummy-pulls.json` file
CleanShot_Google Chrome-Pulldasher

---
### Cominor Pulldasher:
1. Ran `git clone git@github.com:iFixit/pulldasher.git` in my cominor home directory
2. `cd` into the directory and ran `git checkout reaaaaaaaaaaact`
3. Ran `update-pulldasher --run`
4. Went to https://pulldasher.cominor.com/frontend/
5. Noticed that pulls were not being populated either
CleanShot_Google Chrome-Pulldasher- Console 400 Error

**Comment by ardelato** [comment]
dev_block 🐛 

Actually there seems to be a bug that @BaseInfinity stumbled upon.

It seems that the Hovering feature no longer works on the main page.

## Pulldasher on master
https://user-images.githubusercontent.com/22064420/103838218-5a1edb00-5041-11eb-93d9-c03d8e3af5c1.mp4

## Pulldasher updated with this branch
https://user-images.githubusercontent.com/22064420/103838226-6014bc00-5041-11eb-9e8a-69061f184427.mp4

**Comment by danielbeardsley** [comment]
> Confirmation on current working behavior. Should pulls-whether fake or from GitHub-appear on the React Based Pulldasher?

The behavior you witness is the current state of the project (pulls do not appear year). It's a WIP for sure.

dev_block 👍  though on the tooltip bug. We've experienced this when trying to upgrade jQuery. See https://github.com/iFixit/pulldasher/issues/174

**Comment by danielbeardsley** [comment]
I fixed the problem with the tooltips!

See recent commits!

**Comment by danielbeardsley** [comment]
un_dev_block :+1:

**Comment by danielbeardsley** [comment]
image

QA :+1: 

Tooltips on the old UI still work.

The new UI is a WIP and this is a fair chunk of the way there. (At least the hard bootstrapping part is done).

**Comment by danielbeardsley** [comment]
I'm gonna re-mark this as QA :+1:

**Review by andyg0808 — APPROVED** [review]

**Comment by andyg0808** [comment]
CR :fish: through 21d2455

**Comment by andyg0808** [comment]
@danielbeardsley Should we merge this?

**Pull request merged**